yash-builtin/src/kill/syntax.rs

```diff
@@ -151,49 +150,33 @@ impl Error {
 
 /// Converts a string to a signal.
 ///
-/// The string should be a signal name.
-///
-/// If the string is a valid signal name, this function returns `Some(signal)`.
-/// Otherwise, this function returns `None`.
-///
-/// The signal name is parsed case-insensitively.
-///
-/// If `allow_sig_prefix` is `true`, the `SIG` prefix is optional for signal
-/// names. Otherwise, the `SIG` prefix must **not** be present.
-#[must_use]
-pub fn parse_signal_name(s: &str, allow_sig_prefix: bool) -> Option<Signal> {
-    let mut name = String::with_capacity(s.len() + 3);
-    name.push_str(s);
-    name.make_ascii_uppercase();
-    if !allow_sig_prefix || !name.starts_with("SIG") {
-        name.insert_str(0, "SIG");
-    }
-    name.parse().ok()
-}
-
-/// Converts a string to a signal.
-///
-/// The string may be a signal name or a signal number.
+/// The string may be a signal name or a number.
 ///
 /// If the string is a valid signal name or number, this function returns
-/// `Some(Some(signal))`. If the string represents the dummy signal number `0`,
-/// this function returns `Some(None)`. Otherwise, this function returns `None`.
+/// `Some(signal)`. If the string represents the dummy signal number `0`, the
+/// return value will be `Some(Signal::Number(0))`. Otherwise, this function
+/// returns `None`.
 ///
 /// The signal name is parsed case-insensitively.
 ///
 /// If `allow_sig_prefix` is `true`, the `SIG` prefix is optional for signal
 /// names. Otherwise, the `SIG` prefix must **not** be present.
 #[must_use]
-pub fn parse_signal_name_or_number(s: &str, allow_sig_prefix: bool) -> Option<Option<Signal>> {
-    if let Ok(number) = s.parse::<c_int>() {
-        if number == 0 {
-            Some(None)
-        } else {
-            number.try_into().ok().map(Some)
-        }
-    } else {
-        parse_signal_name(s, allow_sig_prefix).map(Some)
+pub fn parse_signal(mut s: &str, allow_sig_prefix: bool) -> Option<Signal> {
+    fn starts_with_sig(s: &str) -> bool {
+        let mut cs = s.chars();
+        matches!(
+            (cs.next(), cs.next(), cs.next()),
+            (Some('S' | 's'), Some('I' | 'i'), Some('G' | 'g'))
+        )
     }
+
+    if allow_sig_prefix && starts_with_sig(s) {
+        // Skip the `SIG` prefix
+        s = &s[3..];
+    }
+
+    s.parse().ok()
```

Comment:
The `SIG` prefix must be matched case-insensitively, that's why I added the `starts_with_sig` function rather than simply using `s.starts_with("SIG")`.

Reply:
Modified the function name for clarity: 4ae6a2c03f9f5416c1c54103eda82b5738888e1b